serving/purgatorio/select-and-train-machine-learning-models/tools-and-libraries/tools-and-libraries.md

```diff
@@ -1,5 +1,321 @@
+
+Every rich Python library contains thousands of different functions and utilities, but _you'll need a small subset to get the things done!_
+
+**So, Virgilio suggests the following approach**:
+
+- Choose a library from the ones that are widely used and feature-rich.
+- Get a general introduction through to Virgilio's resources listed in this guide.
+- Find a problem (similar but not equal to the introductory ones).
+- Dig into the problem, and get your hands dirty with the chosen library.
+- Learn _on the way to the problem solution_.
+- Solve the problem, trying to do it yourself.
+
+If you follow this approach, you take home the following 3 things:
+- **real problem experience**
+- **you feel comfortable coding with a subset of the functions of the tool**
+
+Believe it or not, it works like magic!
+
+Know you know how to solve a new _learning problem_.
+
+Now, how to choose from the myriad of tools and libraries available?
+
+# Navigating the Ocean
+
+It's very easy to find out articles, blog posts, and YouTube videos telling you "the 20 libraries you need to learn in 2020!". 
+
+What? 20 libraries? 
+
+Come on, this is ridiculous _and_ impossible, for a beginner!
+
+Don't even try to do that, or you will miserably get stuck, because it is impossible to learn 20 programming libraries together, even for an expert.
+
+Learning tools and libraries will be a thing most of your life if you want to study or work in the Data Science field. 
+
+**But, as a beginner**:
+
+- is better to start learning very well _some_ tools/libraries, and meanwhile becoming "aware" of the similar ones (and their pro/cons).
+
+- For doing that, it's useful to read the homepage of the official documentation (God bless good documentation!). 
+
+- It's way better to learn **one** (max 2) libraries for each step of the Data Science process, _as a starting point!_
+
+- You'll always be capable of learning more libraries when you reach the capabilities limits of the one that you already know or you have different needs.  
+
+Now you know what is the best approach to learn tools and libraries and frameworks, and how to do it [using wisely your time](https://waitbutwhy.com/2014/05/life-weeks.html)!
+
+The following links expand these points and you should read them if you are a beginner, some answers are illuminating:
+
+- [What are some tips for a beginning ML/data scientist who feels overwhelmed?](https://www.quora.com/What-are-some-tips-for-a-beginning-ML-data-scientist-who-feels-overwhelmed)
+- [SO MUCH to Learn... Overwhelmed. How to Learn so Much?](https://www.reddit.com/r/learnprogramming/comments/dnujl0/so_much_to_learn_overwhelmed_how_to_learn_so_much/)
+- [Beginner to "data science" but unsure of what I should be doing](https://www.reddit.com/r/datascience/comments/7yg61i/beginner_to_data_science_but_unsure_of_what_i/)
+
+Having that said, let's jump in!
```

Comment:
Whenever I speak to people or in my workshops and talks at conferences I mention this:
- identify the nodes of knowledge
- create these nodes of knowledge as we go further
- then traverse slowly deeper into the nodes (branches and sub-branches)
- always start high-level and then go deeper (also called outside-in approach by some methodologies)

I gave a good talk on this long ago, see [slides[(http://bit.ly/java-Nn-know-learn) (also covers how to learn).

Reply:
nice, maybe I can add this in "learning " guide, but it's too deep for this one.